### DISCUSSION TITLE
How exactly does `pyo3-build-config`'s `#[cfg(...)]` macros work?

### DISCUSSION THREAD

**Discussion by jymchng** [PyO3/pyo3#4334] [Questions]
Hi, I have been coding Rust for around two years - I'd also written some procedural macros and a few crates - so I understand how macros work, etc. Thank you for building `pyo3`.

From [this](https://pyo3.rs/main/building-and-distribution/multiple-python-versions#conditional-compilation-for-different-python-versions), I understand how `pyo3` sets certain [`--cfg`](https://blog.rust-lang.org/2024/05/06/check-cfg.html) / [`--check-cfg`](https://blog.rust-lang.org/2024/05/06/check-cfg.html) flags to `rustc`.

If we look at this piece of code copied from the [documentation](https://pyo3.rs/main/building-and-distribution/multiple-python-versions#conditional-compilation-for-different-python-versions):

```
#[cfg(Py_3_7)]
fn function_only_supported_on_python_3_7_and_up() {}
```

The following points can be made:

1. `pyo3` does not control the attribute macro `#[cfg(...)]`, it is controlled by `rustc`.
2. `pyo3` does not control the client's codes, e.g. the function `function_only_supported_on_python_3_7_and_up()`.
3. All `pyo3` has is the flag `Py_3_7`, which in the client's `build.rs` script will be passed to `rustc` and set as a flag.
4. Even if `pyo3` can possibly get the `--cfg` flag (by interrogating `rustc` or through environment variables) - knowing if `Py_3_7` is set; `pyo3` still doesn't control `#[cfg(...)]` so it can't control how the function `function_only_supported_on_python_3_7_and_up()` is expanded.

Hence, my question:

How did `pyo3` accomplish this?
From knowing if a flag, e.g. `Py_3_7`, is set or not to conditional compilation of codes depending on whether the `Python` executable is of version 3.7 and above.

Thank you.

**Comment by jymchng** [comment]
@davidhewitt and others, do you guys have an opinion on this?

**Comment by davidhewitt** [comment]
The cfg flags are set according to the interpreter which `pyo3-build-config` finds when it's being completed. We don't need to change the way `cfg` is expanded.

**Reply by jymchng** [reply]
Consider the example codes from the documentation:

```
#[cfg(Py_3_7)]
fn function_only_supported_on_python_3_7_and_up() {}

#[cfg(not(Py_3_8))]
fn function_only_supported_before_python_3_8() {}

#[cfg(not(Py_LIMITED_API))]
fn function_incompatible_with_abi3_feature() {}
```

If it has detected that the Python execution is of version 3.9, then it will set the flags `Py_3_8` to be `false` and consequently, the second function will not be compiled?

**Reply by davidhewitt** [reply]
Not quite, if Python 3.9 then we set `Py_3_8` too. You should think of `Py_3_8` as meaning "at least Python 3.8".

**Reply by messense** [reply]
See also https://github.com/PyO3/pyo3/blob/a5a3f3f7f28b6a542b29fa552059b8e7bee496fa/pyo3-build-config/src/impl_.rs#L162-L172

**Reply by jymchng** [reply]
> Not quite, if Python 3.9 then we set `Py_3_8` too. You should think of `Py_3_8` as meaning "at least Python 3.8".

Thank you.

In this case, then all 3 functions in the codes example should compile into the binary.

If Python is 3.5, then I guess the second and third function will compile.

If 3.7, then all 3 functions should compile.

**Reply by jymchng** [reply]
> See also
> 
> https://github.com/PyO3/pyo3/blob/a5a3f3f7f28b6a542b29fa552059b8e7bee496fa/pyo3-build-config/src/impl_.rs#L162-L172

Thank you for sharing.
